tests/algorithms/test_algorithm_resumption.py

```diff
@@ -101,157 +26,81 @@
-
-    assert_checkpoints_equivalent(hparams_a=trainer_hparams,
-                                  checkpoint_file_a=first_trainer_final_checkpoint_filepath,
-                                  hparams_b=second_trainer_hparams,
-                                  checkpoint_file_b=second_trainer_final_checkpoint_filepath,
-                                  state_attrs_to_skip=[
-                                      '_optimizers',
-                                      '_schedulers',
-                                      '_algorithms',
-                                      '_callbacks',
-                                  ])
+    if algorithm in ('no_op_model', 'scale_schedule'):
+        pytest.skip('stub algorithms')
+
+    if algorithm in ('cutmix, mixup, label_smoothing'):
+        # see: https://github.com/mosaicml/composer/issues/362
+        pytest.importorskip("torch", minversion="1.10", reason="Pytorch 1.10 required.")
+
+    if algorithm in ('layer_freezing', 'swa', 'stochastic_depth'):
+        pytest.xfail('Known issues')
+
+    setting = get_settings(algorithm)
+    if setting is None:
+        pytest.skip('No setting provided in algorithm_settings.')
+
+    folder1 = os.path.join(tmpdir, 'folder1')
+    folder2 = os.path.join(tmpdir, 'folder2')
+
+    config = {
+        'algorithms': setting['algorithm'],
+        'model': setting['model'],
+        'train_dataloader': DataLoader(dataset=setting['dataset'], batch_size=4),
+        'max_duration': '5ep',
+        'loggers': [],
+        'seed': seed,
+        'device': device,
+        'deterministic_mode': True,
+        'save_filename': save_filename,
+        'save_folder': folder1,
+        'save_interval': save_interval,
+        'train_subset_num_batches': 5,
+    }
 
-    # Clean up passed checkpoints
-    tmpdir.remove()
+    # train model once, saving checkpoints every epoch
+    trainer1 = Trainer(**config)
+    trainer1.fit()
+
+    # create second trainer, load an intermediate checkpoint
+    # and continue training
+    setting = get_settings(algorithm)
+    assert setting is not None
+
+    config.update({
+        'model': setting['model'],
+        'save_folder': folder2,
+        'load_path': os.path.join(folder1, resume_file),
+        'load_weights_only': False,
+        'load_strict': False,
+    })
+    trainer2 = Trainer(**config)
+    trainer2.fit()
+
+    # check that the checkpoints are equal
+    _assert_checkpoints_equal(
+        file1=os.path.join(folder1, final_checkpoint.format(rank=0)),
+        file2=os.path.join(folder2, final_checkpoint.format(rank=0)),
+    )
+
+
+def _assert_checkpoints_equal(file1, file2):
+    checkpoint1 = torch.load(file1)
+    checkpoint2 = torch.load(file2)
+
+    # compare rng
+    deep_compare(checkpoint1['rng'], checkpoint2['rng'])
+
+    # compare state
+    _check_dict_recursively(checkpoint1['state'], checkpoint2['state'])
```

Comment:
`tests.common` includes a `assert_state_equivalent`.

(This function should probably be split up into `assert_state_equivalent` and `assert_state_dict_equivalent`)



```suggestion
    assert_state_dict_equivalent(checkpoint1['state'], checkpoint2['state'])
```

Reply:
I switched this to use `deep_compare`, which seems to pass the tests with `atol=0, rtol=0` on GPU. This is stricter than the `assert_state_dict_equivalent`, which has a pretty worrisome tolerances:

```
    atol = 0.1 if is_gpu else 0.0
    rtol = 0.1 if is_gpu else 0.0
```
FYI, we've seen very reproducible ResNet50 runs, where the same seed yields essentially the same accuracy over 90 epochs of training, so we should definitely trace where this is coming from.